### PR TITLE
[tplinksmarthome] Added support for power outlets HS107, HS300, KP200, KP400

### LINKED ISSUE #5051: [tplinksmarthome] Add support for TP-Link HS107, HS300, KP200, and KP400
HS107, HS300, KP200, and KP400 are devices with more than 1 outlet on the same device. The HS107, KP200 and KP400 all have 2, and the HS300 has 6 outlets. These outlets can be controlled separately and also have the ability to read power usage. Because of this they also have different control commands and require additional programming to be supported by the tplinksmarthome binding.

_This issue report is also to report some work already has been done to add support for these devices. However since not all commands are known it's still a work in progress. And since I don't have either device I can't test this._

Anyone who has such a device and is interested in support for one of these devices please comment in this item.

### Power outlet devices:
* HS107 Smart Wi-Fi Plug, 2-Outlets
* HS300 Smart Wi-Fi Power Strip
* KP200 Smart Wi-Fi Power Outlet, 2-Sockets
* KP400 Smart Outdoor Plug

### Open questions
The following questions remain open with regards to how the devices work:
1. Is it possible to switch all outlets at once?
1. Energy reading command needs to be given for each outlet (based on what I see with other implementations of support for these devices). Is is possible to get readings for multiple devices at once?
1. Is is possible to get a total energy reading for all outlets combined?
1. Can the status leds be switch on and off like with the plugs and switched?

### PR DESCRIPTION
**Pull request by Hilbrand**
Closes #5051

### CODE REVIEW

File: bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/TPLinkSmartHomeDiscoveryService.java [hunk 1]
```diff
@@ -165,24 +166,25 @@ private void closeDiscoverSocket() {
...
+        final Sysinfo sysinfoRaw = commands.getSysinfoReponse(rawData);
+        final Sysinfo sysinfo = sysinfoRaw.getActualSysinfo();
 
         logger.trace("Detected TP-Link Smart Home device: {}", rawData);
-        String deviceId = sysinfo.getDeviceId();
+        final String deviceId = sysinfo.getDeviceId();
         logger.debug("TP-Link Smart Home device '{}' with id {} found on {} ", sysinfo.getAlias(), deviceId, ipAddress);
         idInetAddressCache.put(deviceId, ipAddress);
-        Optional<ThingTypeUID> thingTypeUID = getThingTypeUID(sysinfo.getModel());
+        final Optional<ThingTypeUID> thingTypeUID = getThingTypeUID(sysinfo.getModel());
```
Comment (martinvw) [inline bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/TPLinkSmartHomeDiscoveryService.java]: Still can't get used to it...

File: bundles/org.openhab.binding.tplinksmarthome/src/main/resources/ESH-INF/thing/HS300.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            </channel-group>
+            <channel-group id="outlet5" typeId="energy-outlet" >
+                <label>Outlet 5</label>
+            </channel-group>
+            <channel-group id="outlet6" typeId="energy-outlet" >
+                <label>Outlet 6</label>
+            </channel-group>
+        </channel-groups>
+
+		<representation-property>deviceId</representation-property>
```
Comment (martinvw) [inline bundles/org.openhab.binding.tplinksmarthome/src/main/resources/ESH-INF/thing/HS300.xml]: Looks like a mix of whitespace

File: bundles/org.openhab.binding.tplinksmarthome/src/main/resources/ESH-INF/thing/channels.xml [hunk 3]
```diff
@@ -58,4 +58,32 @@
...
+	<channel-group-type id="switch-outlet">
+		<label>Outlet</label>
+		<category>PowerOutlet</category>
+		<channels>
+			<channel id="switch" typeId="system.power" />
+		</channels>
+	</channel-group-type>
+	<channel-group-type id="energy-outlet">
+		<label>Outlet</label>
+        <category>PowerOutlet</category>
```
Comment (martinvw) [inline bundles/org.openhab.binding.tplinksmarthome/src/main/resources/ESH-INF/thing/channels.xml]: Mix of whitespace

File: bundles/org.openhab.binding.tplinksmarthome/src/test/java/org/openhab/binding/tplinksmarthome/internal/device/PowerStripDeviceTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
```
Comment (martinvw) [inline bundles/org.openhab.binding.tplinksmarthome/src/test/java/org/openhab/binding/tplinksmarthome/internal/device/PowerStripDeviceTest.java]: This headers looks different / wrong

File: bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/device/BulbDevice.java [hunk 5]
```diff
@@ -103,7 +103,7 @@ private HasErrorResponse handleDecimalType(String channelID, Connection connecti
         return null;
     }
 
-    private TransitionLightStateResponse handleColorTemperature(Connection connection, int colorTemperature,
+    private @Nullable TransitionLightStateResponse handleColorTemperature(Connection connection, int colorTemperature,
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/device/BulbDevice.java]: As the `connection` property has been moved to the super class `SmartHomeDevice` can it be removed here - and elsewhere?

File: bundles/org.openhab.binding.tplinksmarthome/README.md [hunk 6]
```diff
@@ -241,33 +254,44 @@ Either `deviceId` or `ipAddress` must be set.
...
+|------------------|-----------|----------------------------------------------------|---------------------------------------------------------------------------------------------|
+| switch           | Switch    | Switch the Smart Home device on or off.            | HS100, HS103, HS105, HS107, HS110, HS200, HS210, HS300, KP100, KP200, KP400, RE270K, RE370K |
+| brightness       | Dimmer    | Set the brightness of Smart Home device or dimmer. | HS220, KB100, KL110, KL120, LB100, LB110, LB120, LB200                                      |
+| colorTemperature | Dimmer    | Set the color temperature of Smart Home light.     | KB130, KL120, KL130, LB120, LB130, LB230                                                    |
+| color            | Color     | Set the color of the Smart Home light.             | KB130, KL130, LB130, LB230                                                                  |
+| power            | Number    | Actual energy usage in Watt.                       | HS110, HS300, KLxxx, LBxxx                                                                  |
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.tplinksmarthome/README.md]: A second question: Why don't you use UoM here. Are there plans to introduce it for this binding.

File: bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/device/BulbDevice.java (hunk 5 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/device/BulbDevice.java]: Yes makes sense. I've removed the `connection` and `configuration` parameters.

Comment (Hilbrand) [reply bundles/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/device/BulbDevice.java]: ok got merged before I could add :smile: (I've create a new pr: #5766)

File: bundles/org.openhab.binding.tplinksmarthome/README.md (hunk 6 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.tplinksmarthome/README.md]: I was wondering about that too recently :smile: The original implementation (for HS110) is from before UoM and the others just reuse the code. I was thinking about making them UoM, although it's also a breaking change for the HS110. But I was planning to change it after this pr was being merged.

**Pull request merged**